runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java

```diff
@@ -166,6 +222,134 @@ public void onOverheadTextChanged(final OverheadTextChanged event)
 		event.getActor().setOverheadText(updatedMessage);
 	}
 
+	@Subscribe
+	public void onMenuEntryAdded(MenuEntryAdded event)
+	{
+		int childId = WidgetInfo.TO_CHILD(event.getActionParam1());
+		if (!config.showEmitEmojiEntry() || childId != WidgetInfo.RESIZABLE_VIEWPORT_ALL_CHAT_TAB.getChildId())
+		{
+			return;
+		}
+
+		MenuEntry[] entries = client.getMenuEntries();
+		for (MenuEntry entry : entries)
+		{
+			if (entry.getOption().equals(EMIT) && entry.getTarget().equals(MENU_TARGET))
+			{
+				return;
+			}
+		}
+		final MenuEntry emitEmojis = new MenuEntry();
+		emitEmojis.setOption(EMIT);
+		emitEmojis.setTarget(MENU_TARGET);
+		emitEmojis.setIdentifier(event.getIdentifier());
+		emitEmojis.setParam0(event.getActionParam0());
+		emitEmojis.setParam1(event.getActionParam1());
+		emitEmojis.setType(MenuAction.RUNELITE_EMOJI_OUTPUT.getId());
+		insertMenuEntry(emitEmojis, client.getMenuEntries());
+	}
+
+	@Subscribe
+	public void onGameTick(GameTick gameTick)
+	{
+		timeoutEmojiMessages();
+	}
+
+	@Subscribe
+	public void onMenuOptionClicked(MenuOptionClicked event)
+	{
+		if (event.getMenuAction().equals(MenuAction.RUNELITE_EMOJI_OUTPUT))
+		{
+			addEmojiPaletteMessage();
+		}
+	}
+
+	private void insertMenuEntry(MenuEntry newEntry, MenuEntry[] entries)
+	{
+		MenuEntry[] newMenu = ObjectArrays.concat(entries, newEntry);
+		int menuEntryCount = newMenu.length;
+		// swap Cancel entry to the end of the array
+		ArrayUtils.swap(newMenu, menuEntryCount - 1, menuEntryCount - 2);
+		client.setMenuEntries(newMenu);
+	}
+
+	private void timeoutEmojiMessages()
+	{
+		if (emojiEmitMessages.isEmpty())
+		{
+			return;
+		}
+
+		boolean removed = false;
+
+		for (Iterator<EmojiEmitMessage> it = emojiEmitMessages.iterator(); it.hasNext(); )
+		{
+			EmojiEmitMessage emojiEmitMessage = it.next();
+			MessageNode messageNode = emojiEmitMessage.getMessageNode();
+			final int createdTick = emojiEmitMessage.getTick();
+
+			if (client.getTickCount() > createdTick + EMOJI_EMIT_DURATION)
+			{
+				it.remove();
+
+				// If this message has been reused since, it will get a different id
+				if (emojiEmitMessage.getGetMessageId() == messageNode.getId())
+				{
+					ChatLineBuffer ccInfoBuffer = client.getChatLineMap().get(ChatMessageType.GAMEMESSAGE.getType());
+					if (ccInfoBuffer != null)
+					{
+						ccInfoBuffer.removeMessageNode(messageNode);
+						removed = true;
+					}
+				}
+			}
+			else
+			{
+				// Everything else in the deque is newer
+				break;
+			}
+		}
+
+		if (removed)
+		{
+			clientThread.invoke(() -> client.runScript(ScriptID.BUILD_CHATBOX));
+		}
+	}
+
+	private ChatMessageBuilder getEmojiMessageBuilder()
+	{
+		ChatMessageBuilder builder = new ChatMessageBuilder();
+		Emoji[] allEmoji = getAllEmoji();
+
+		for (int i = 0; i < allEmoji.length; i++)
+		{
+			Emoji emoji = allEmoji[i];
+			builder
+				.img(modIconsStart + emoji.ordinal())
+				.append("  " + Text.unescapeTags(emoji.trigger) + "  ");
```

Comment:
Maybe something like `FontTypeFace#getTextWidth` can be used when aligning these(?)
The current formatting unfortunately looks really messy.

Reply:
Yup, I completely agree. I updated the PR description with the output as it stands with the latest changes.